File: simple-clojure-http-server/src/simple_clojure_http_server/request_handler.clj
(ns simple-clojure-http-server.request-handler
  (:require [clojure.java.io :as io])
  (:import (java.nio.file Files LinkOption Paths)))

(defn- file->bytes [path]
  (with-open [xin (io/input-stream path)
              xout (java.io.ByteArrayOutputStream.)]
    (io/copy xin xout)
    (.toByteArray xout)))

(defn- is-file-exists? [path]
  (.exists (io/file path)))

(defn- is-directory? [path]
  (Files/isDirectory
   (.normalize
    (Paths/get "public" (into-array [path])))
   (into-array LinkOption nil)))

(defn normalize [path]
  (if (is-directory? path)
    "public/index.html"
    (str "public" path)))

(defn handle-not-found []
  {:status 404
   :body (file->bytes "public/404.html")})

(defn handle-ok [path]
  {:status 200
   :body (file->bytes path)})

(defn handle [{:keys [method path version]}]
  (let [normalized-path (normalize path)]
    (if (is-file-exists? normalized-path)
      (handle-ok normalized-path)
      (handle-not-found))))

File: simple-kotlin-http-server/src/main/kotlin/com/whammy/RequestHandler.kt
package com.whammy

import java.nio.file.Files
import java.nio.file.Paths

class Request(val method: String, val path: String, val version: String)

class RequestHandler {

    private val contentType = "text/html;charset=utf8"

    fun handleRequest(request: Request): Response {
        val path = Paths.get("public", request.path)
        println(path.toAbsolutePath().toString())
        if (!path.startsWith("public/")) {
            return Response(403, contentType, Files.readAllBytes(Paths.get("public", "403.html")))
        }
        else if (Files.isRegularFile(path) && Files.exists(path)) {
            return Response(200, contentType, Files.readAllBytes(path))
        }
        else if (Files.isDirectory(path) && Files.exists(path.resolve("index.html"))) {
            return Response(200, contentType, Files.readAllBytes(Paths.get("public", "index.html")))
        }
        return Response(404, contentType, Files.readAllBytes(Paths.get("public", "404.html")))
    }
}